app/src/main/java/org/oppia/app/player/state/itemviewmodel/FractionInteractionViewModel.kt

```diff
@@ -27,6 +33,46 @@ class FractionInteractionViewModel(
     return userAnswerBuilder.build()
   }
 
+  override fun isExplicitErrorCheckRequired(): Boolean = true
+
+  override fun getPendingAnswerError(category: AnswerErrorCategory): String? {
+    if (category == AnswerErrorCategory.REAL_TIME) {
+      return if (answerText.isNotEmpty() && StringToFractionParser().getRealTimeError(answerText.toString()) != FractionParsingError.VALID)
+        StringToFractionParser().getRealTimeError(answerText.toString()).getErrorMessageFromStringRes(context)
+      else
+        null
+    } else
+      return getPendingAnswerErrorOnSubmit()
+  }
+
+  fun getPendingAnswerErrorOnSubmit(): String? {
+    return if (answerText.isNotEmpty() && StringToFractionParser().getSubmitTimeError(answerText.toString()) != FractionParsingError.VALID) {
+      errorMessage.set(
```

Comment:
It's suspect to me that we're setting the error & retrieving it in the same place--getters generally don't have side effects because that can lead to unintended consequences. Can we separate this out? E.g. I'd actually expect the submission flow or the text watcher to set the error, and these getter methods just retrieve the correct error at the correct time.

Reply:
Done but iam confused where you want me to use this getter because on both during the submission flow and the text watcher we set the error

Reply:
@BenHenning but iam confused where you want me to use this getter because during both on submission and the text watcher we set the error

Reply:
Thanks. This is closer to what I was thinking, but three aspects confusing me:
1) The submit button should also be setting an error in certain situations (this doesn't seem to be the case currently--only the keyboard button is doing this)
2) I'd expect that the TextWatcher sends a real-time error event, and then the presenter uses that to set the error. In this, only two places ever set the error message and both are in the state fragment presenter: one for all real-time errors (triggered by text watcher for fractions), and submit-time errors (triggered whenever the user tries to submit an answer).
3) No error message setting should happen within the view model, which I think in turn simplifies the view model a bunch

Does this help?

Reply:
Done.